Refactor Rigid Bodies Body Definition

b-butler opened the pull request:
## Description
Remove charge/diameter from rigid body definition. Allows setting charges in createRigidBodies, but not diameters since we are removing that.

## Motivation and context
This is one of the main tasks for 4.0.

Resolves #1350

## How has this been tested?
Existed _modified_ tests are sufficient.

## Change log

<!-- Propose a change log entry. -->
```
Changed: `hoomd.md.constrain.Rigid` no longer takes `diameters` or `charges` as body arguments. The `create_bodies` method now takes an optional `charges` argument to set charges.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

b-butler committed:
refactor!: Remove charge/diameter from rigid body definition.

Allows setting charges in createRigidBodies, but not diameters since we
are removing that.

b-butler committed:
fix: Remove diameter from rigid body tests

b-butler commented:
@joaander the containers are failing to initialize which results in test failing.

It looks like the GPU driver is causing the error?

joaander commented:
@b-butler the tests are now running.

joaander commented on hoomd/md/ForceComposite.cc:
```diff
@@ -8,6 +8,10 @@
 
+#ifndef __HIPCC__
+#include <pybind11/stl.h>
+#endif
```
```suggestion
#include <pybind11/stl.h>
```
Check not needed in `.cc` files.

joaander commented on hoomd/md/ForceComposite.cc:
```diff
@@ -513,7 +489,39 @@ void ForceComposite::validateRigidBodies()
+        std::vector<float> charges_vector;
+        for (auto& charge : charges_list)
+            {
+            charges_vector.emplace_back(charge.cast<float>());
```
```suggestion
            charges_vector.emplace_back(charge.cast<Scalar>());
```

joaander commented on hoomd/md/ForceComposite.cc:
```diff
@@ -513,7 +489,39 @@ void ForceComposite::validateRigidBodies()
+            {
+            throw std::runtime_error("Charges provided not consistent with rigid body size.");
+            }
+        std::vector<float> charges_vector;
```
```suggestion
        std::vector<Scalar> charges_vector;
```

joaander commented on hoomd/md/constrain.py:
```diff
@@ -290,19 +290,20 @@ def __init__(self):
 
         Args:
             state (hoomd.State): The state in which to create rigid bodies.
+            charges (dict[str, float], optional): The charges for each of the
```
```suggestion
            charges (dict[str, list[float]], optional): The charges for each of the
```

joaander reviewed (changes requested):
Thanks! This turned out to be a minimal change to fix the API.

Update `test_rigid_filter` to work with this change: https://github.com/glotzerlab/hoomd-blue/actions/runs/3438664815/jobs/5803196403#step:6:4513

b-butler committed:
refactor: Remove unnecessary ifdef guard

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

b-butler committed:
doc: Fix type documentation for Rigid.create_bodies

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

b-butler committed:
fix: Use Scalar typedef for charges in createRigidBodies

joaander commented:
`test_filter_md.py` is still failing.

b-butler commented:
Sorry missed that.

joaander commented:
We also need to make a `trunk-major` branch on https://github.com/glotzerlab/hoomd-examples, set it to build the `trunk-major` branch of HOOMD for testing, and update the rigid body tutorial to match the new API.

b-butler committed:
test: Fix hoomd.filter.Rigid tests.

Needed to fix rigid body definition.

joaander commented on hoomd/md/constrain.py:
```diff
@@ -290,19 +290,20 @@ def __init__(self):
 
         Args:
             state (hoomd.State): The state in which to create rigid bodies.
+            charges (dict[str, list[float]], optional): The charges for each of the
+                constituent particles, defaults to ``None``. If ``None``, all
+                charges are zero. The keys should be the central particles.
```
```suggestion
            charges (dict[str, list[float]], optional): The charges for each of
                the constituent particles, defaults to ``None``. If ``None``,
                all charges are zero. The keys should be the central particles.
```
fix: `hoomd/md/constrain.py:304:81: E501 line too long (83 > 80 characters)`

joaander reviewed (approved):
Thanks! This looks good.

One small change is needed so that flake8 will pass.

joaander committed:
Update hoomd/md/constrain.py

b-butler committed:
fix: Correctly set constituent particles pos in create bodies

Need to place constituent particles where they will be decomposed to the
same rank.

joaander commented:
Reminder to update the tutorial: https://github.com/glotzerlab/hoomd-examples/issues/72
